classes/property_list.php

```diff
@@ -0,0 +1,346 @@
+     * @throws \DOMException
+     */
+    public function __construct(string $xml = '') {
+        $this->cfpropertylist = new CFPropertyList();
+
+        if (empty($xml)) {
+            // If xml not provided, create a blank PList with root dictionary set up.
+            $this->cfpropertylist->add(new CFDictionary([]));
+        } else {
+            // Parse the XML into a PList object.
+            $this->cfpropertylist->parse($xml, CFPropertyList::FORMAT_XML);
+        }
+    }
+
+    /**
+     * Add a new element to the root dictionary element.
+     *
+     * @param string $key Key to assign to new element.
+     * @param CFType $element The new element. May be a collection such as an array.
+     */
+    public function add_element_to_root(string $key, CFType $element) {
+        // Get the PList's root dictionary and add new element.
+        $this->cfpropertylist->getValue()->add($key, $element);
+    }
+
+    /**
+     * Get value of element identified by key.
+     *
+     * @param string $key Key of element.
+     * @return mixed Value of element found, or null if none found.
+     */
+    public function get_element_value(string $key) {
+        $result = null;
+        $this->plist_map( function($elvalue, $elkey, $parent) use ($key, &$result) {
+            // Convert date to iso 8601 if date object.
+            if ($key === $elkey) {
+                $result = $elvalue->getValue();
+            }
+        }, $this->cfpropertylist->getValue());
+
+        if (is_array($result)) {
+            // Turn CFType elements in PHP elements.
+            $result = $this->array_serialize_cftypes($result);
+        }
+        return $result;
+    }
+
+    /**
+     * Update the value of any element with matching key.
+     *
+     * Only allow string, number and boolean elements to be updated.
+     *
+     * @param string $key Key of element to update.
+     * @param mixed $value  Value to update element with.
+     */
+    public function update_element_value(string $key, $value) {
+        if (is_array($value)) {
+            debugging('property_list: Can\'t call update_element_value with array. Use update_element_array instead.',
+                    DEBUG_DEVELOPER);
+            return;
+        }
+        $this->plist_map( function($elvalue, $elkey, $parent) use ($key, $value) {
+            // Set new value.
+            if ($key === $elkey) {
+                $element = $parent->get($elkey);
+                // Limit update to boolean and strings types, and check value matches expected type.
+                if (($element instanceof CFString && is_string($value))
+                        || ($element instanceof CFNumber && is_numeric($value))
+                        || ($element instanceof CFBoolean && is_bool($value))) {
+                    $element->setValue($value);
+                } else {
+                    debugging('property_list: Can only update strings, numbers and bools with corresponding type.',
```

Comment:
Should we terminate and throw an exception and not silently swallow broken value?

Reply:
Replaced debugging with exceptions for bad params if updating. Adjusted unit tests to expect exceptions instead of debugging.